### PR TITLE
Nested Arc - Envoy proxy and DNS customization for nested comms

### PR DESCRIPTION
**Pull request by katriendg**
- Use Helm charts for Envoy and DNSMasq to deploy these components upon infrastructure bootstrapping
- Adapt CoreDNS per cluster to rewrite allow-domains passing through Envoy proxy
- Configure DNS resolver from lower layers to upper layer
- Adpat NSG rules to Close off outbound Internet access for lower layer clusters (changes to bicep and order of execution of these)
- Bootstrap AKS cluster and Azure Arc agent through the proxy
- Reverse Proxy documentation

### CODE REVIEW

File: docs/reverseproxy.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+For modifying some of the domain names that must pass through the Envoy proxy, the CoreDNS plugin model is employed. Customization can be done using the `hosts` and custom `forward` servers.
+
+Upon a successful installation of the Envoy proxy, a load balancer service endpoint listening on port 443 is generated. This internal endpoint is assigned to a set of domain names to be overwritten via the `hosts` and `rewrite` plugin model for CoreDNS. In Azure AKS, this customization is possible but limited, and it can be accomplished through the ConfigMap `coredns-custom`. Please see [Customize CoreDNS with Azure Kubernetes Service](https://learn.microsoft.com/en-us/azure/aks/coredns-custom) for more information.
+
+## Virtual Network Configuration
+
+### Inbound (Ingress) Network Configuration
+
+At the networking level within the Virtual Network, all traffic from lower layers in a peered VNET is allowed through. This can further be locked down and in this case NSGs should permit traffic inbound to upper levels for ports 443 (HTTPS), 8883 (MQTT), 8084 (for Arc) and 53 (custom DNS).
```
Comment (suneetnangia) [inline docs/reverseproxy.md:43]: Not sure, did add a comment re allowing all FQDN resolutions via DNS at top level being less ideal but allowed in sample?

**Review by suneetnangia — COMMENTED** [review]
What a significant amount of challenges addressed in this PR. Thank you!

File: docs/reverseproxy.md (hunk 1 above)
Comment (katriendg) [reply docs/reverseproxy.md:43]: Do you mean with L4 having full Internet access and using the default Azure DNS resolution? There is a note about potential usage of an enterprise firewall vs full Internet access, but DNS resolution for public FQDN will typically be at the enterprise's level DNS which may be pointing to a public DNS. 

There is a mention of usage of DNSMasq as a non-production sample, but I think you mean something else with your comment. 

Where do you suggest adding the note?

Comment (suneetnangia) [reply docs/reverseproxy.md:43]: Yeah, I believe in the current implementation, all FQDNs will be resolved from all layers regardless of whether those are permitted to connect via proxy or not (that's impression I had from our discussion but please correct). A note highlighting this as a potential breach of Principle of Least Privilege might be useful so users can lock DNS to only permissible FQDNs?

Comment (katriendg) [reply docs/reverseproxy.md:43]: Yes you are right, got it now and remember our discussion. I've added a paragraph about default resolution for all FQDNS and how this should be prevented in production scenario.

File: docs/reverseproxy.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+By updating the Virtual Network DNS setting for L2 and L3 to each use the upstream DNSMasq service, the host system will be able to resolve the URLs that are allowed to pass through to the Internet. 
+
+> It is also worth noting that L2 and L3 DNS resolution for FQDNs that are not configured in DNSMasq will be resolved by the default Azure DNS service, see [What is IP address 168.63.129.16](https://learn.microsoft.com/en-us/azure/virtual-network/what-is-ip-address-168-63-129-16). This means that even FQDNs that are not in the allow-list going through Envoy proxy can be resolved. The requests to these URLs will never work because outbound traffic is blocked by default. This is a solution that is acceptable for this sample but in a production environment one would want to make sure only allow-list resolution is able to be resolved, and not all public FQDNs, regardless of whether the requests would get through or not.
```
[the remark this answers is not in the document]
Comment (suneetnangia) [reply docs/reverseproxy.md:70]: LGTM. Thanks.

**Review by suneetnangia — APPROVED** [review]

**Pull request merged**